### PR TITLE
Add open api doc generation goal

### PR DESCRIPTION
**Pull request by loosebazooka**

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          "-w", webappDir.getPath()));
+      if (!Strings.isNullOrEmpty(hostname)) {
+        params.add("-h");
+        params.add(hostname);
+      }
+      if (serviceClasses != null) {
+        params.addAll(serviceClasses);
+      }
+
+      System.out.print(params);
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: This should be a debug level log message.

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  /**
+   * Default hostname of the Endpoint Host.
+   */
+  @Parameter(property = "endpoints.hostname", required = false)
+  private String hostname;
+
+  public void execute() throws MojoExecutionException {
+    try {
+      if (!openApiDocDir.exists()) {
+        if (!openApiDocDir.mkdirs()) {
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: You can combine these `if`s.

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      }
+      if (serviceClasses != null) {
+        params.addAll(serviceClasses);
+      }
+
+      System.out.print(params);
+      new EndpointsTool().execute(params.toArray(new String[params.size()]));
+
+
+    } catch (Exception e) {
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: Can this be a more specific exception class?

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      throw new MojoExecutionException("Endpoints Tool Error", e);
+    }
+  }
+
+  private String computeOpenApiDocPath() {
+    return new File(openApiDocDir, "openapi.json").getPath();
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: Should this use `getAbsolutePath()`?

File: src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java [hunk 5]
```diff
@@ -45,38 +45,43 @@
...
-
-    MavenXpp3Reader reader = new MavenXpp3Reader();
-    Model model = reader.read(new FileReader("pom.xml"));
-
-    pluginVersion = model.getVersion();
   }
 
+
   @Test
   public void testServerArtifactCreation() throws IOException, VerificationException {
```
Comment (akerekes) [inline src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java]: This test should be split into 3 separate tests. Probably in a separate PR.

File: src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java [hunk 6]
```diff
@@ -87,7 +92,8 @@ public void testServerArtifactCreation() throws IOException, VerificationExcepti
     String discovery = Files.toString(new File(testDir, DISC_DOC_PATH), Charsets.UTF_8);
     Assert.assertThat(discovery, JUnitMatchers.containsString(DEFAULT_URL));
 
-
+    String openapi = Files.toString(new File(testDir, OPEN_API_DOC_PATH), Charsets.UTF_8);
+    Assert.assertThat(openapi, JUnitMatchers.containsString(DEFAULT_HOSTNAME));
   }
 
   @Test
```
Comment (akerekes) [inline src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java]: This test should be split into 3 separate tests. Probably in a separate PR.

File: src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java [hunk 7]
```diff
@@ -105,6 +111,10 @@ public void testApplicationId() throws IOException, VerificationException {
     String discovery = Files.toString(new File(testDir, DISC_DOC_PATH), Charsets.UTF_8);
     Assert.assertThat(discovery, CoreMatchers.not(JUnitMatchers.containsString(DEFAULT_URL)));
     Assert.assertThat(discovery, JUnitMatchers.containsString("https://maven-test.appspot.com/_ah/api"));
+
+    String openapi = Files.toString(new File(testDir, OPEN_API_DOC_PATH), Charsets.UTF_8);
+    Assert.assertThat(discovery, CoreMatchers.not(JUnitMatchers.containsString(DEFAULT_HOSTNAME)));
+    Assert.assertThat(openapi, JUnitMatchers.containsString("maven-test.appspot.com"));
   }
 
   @Test
```
Comment (akerekes) [inline src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java]: This test should be split into 3 separate tests. Probably in a separate PR.

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.Execute;
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.plugins.annotations.ResolutionScope;
+import org.apache.maven.project.MavenProject;
+
+/**
+ * Goal which generates discovery docs
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: `openapi`

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+ */
+@Mojo(name = "openApiDocs", requiresDependencyResolution = ResolutionScope.COMPILE)
+@Execute(phase = LifecyclePhase.PREPARE_PACKAGE)
+public class OpenApiDocsMojo extends AbstractEndpointsWebAppMojo {
+
+  @Parameter(defaultValue = "${project}", readonly = true)
+  private MavenProject project;
+
+  /**
+   * Output directory for discovery docs
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: `openapi`

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+   */
+  @Parameter(defaultValue = "${project.build.directory}/openapi-docs",
+             property = "endpoints.openApiDocDir", required = true)
+  private File openApiDocDir;
+
+  /**
+   * Default hostname of the Endpoint Host.
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: Does Endpoint Host have to be capitalized?

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+   */
+  @Parameter(property = "endpoints.hostname", required = false)
+  private String hostname;
+
+  public void execute() throws MojoExecutionException {
+    try {
+      if (!openApiDocDir.exists()) {
+        if (!openApiDocDir.mkdirs()) {
+          throw new MojoExecutionException(
+              "Failed to create output directory: " + openApiDocDir.getPath());
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: `getAbsolutePath()`?

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java (hunk 10 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: probably not.

File: src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java (hunk 5 above)
Comment (loosebazooka) [inline src/test/java/com/google/cloud/tools/maven/endpoints/framework/ProjectTest.java]: #21

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java (hunk 2 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: I think I need to them to be separate to fail if the directory fails to be created.

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java (hunk 3 above)
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/OpenApiDocsMojo.java]: EndpointsTools throws Exception

**Comment by loosebazooka** [comment]
@akerekes PTAL

File: src/main/java/com/google/cloud/tools/maven/endpoints/framework/EndpointsGenSrcMojo.java [hunk 12]
```diff
@@ -52,7 +52,8 @@
   /**
    * Output directory for generated sources
    */
-  @Parameter(defaultValue = "${project.build.directory}/generated-sources/endpoints",
+  //@Parameter(defaultValue = "${project.build.directory}/generated-sources/endpoints",
```
Comment (akerekes) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/EndpointsGenSrcMojo.java]: Revert this.

Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/EndpointsGenSrcMojo.java]: Why did I even make that :tractor:

Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/maven/endpoints/framework/EndpointsGenSrcMojo.java]: oh the change didn't show the change fully.

**Comment by patflynn** [comment]
For this kind of PR it would have been nice to have a linked issue, or at least a description of the surface change in the PR details. We need this to have useful release notes.

**Pull request merged**